### ISSUE TITLE
[BUG] cuDNN error when training parametric LSTM model

### ISSUE THREAD

**Issue by satellitnorden** [sdatkinson/neural-amp-modeler#450]
**Description**
I used the old parametric code and tried to train a parametric LSTM model, but get the following error:
"RuntimeError: cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input."
I do not get this error when training a WaveNet model. I suspect there might be a bug with how we feed/mix in the parametric data into the inputs of the training process.

**To Reproduce**
- Re-enable/sync to the old parametric code.
- Set up a parametric LSTM model with at least 1 parameter.
- Try to train.
- Observe error.

**Expected behavior**
The model should start training as expected.

**Desktop (please complete the following information):**
 - OS Windows 11
 - Local CLI trainer.
 - Commit hash: 6d33b41cf9f361b774e25c4e1e6340503ae7410e

**Additional Context**
If I put a `torch.backends.cudnn.enabled = False` at the beginning of `main` in `full.py`, it seems like it is training as it should, although painfully slow as I suspect that disables GPU training.

**Comment by prash-p** [comment]
I'm also having this issue with any LSTM model I try to train. I'd rather not disable cudnn

**Comment by sdatkinson** [comment]
Interesting:
```
conda create -n test python  # I'm getting Python 3.12.4
conda activate test
# From https://pytorch.org/:
conda install pytorch pytorch-cuda=12.4 -c pytorch -c nvidia -y  # PyTorch 2.4.0
conda install numpy -y
conda list  # Output below
python
>>> import torch
>>> assert torch.cuda.is_available()
>>> input_size=5
>>> x = torch.randn(2, 3, input_size).to("cuda")
>>> model = torch.nn.LSTM(num_layers=1, input_size=input_size, hidden_size=12).to("cuda")
>>> model(x)  # Error?...Nope, this succeeds!
```

So there is indeed something that's going wrong in between PyTorch and NAM.

Output of `conda list`:
```
# packages in environment at C:\Users\steve\anaconda3\envs\test:
#
# Name                    Version                   Build  Channel
blas                      1.0                         mkl
bzip2                     1.0.8                h2bbff1b_6
ca-certificates           2024.7.2             haa95532_0
cuda-cccl                 12.6.37                       0    nvidia
cuda-cccl_win-64          12.6.37                       0    nvidia
cuda-cudart               12.4.127                      0    nvidia
cuda-cudart-dev           12.4.127                      0    nvidia
cuda-cupti                12.4.127                      0    nvidia
cuda-libraries            12.4.0                        0    nvidia
cuda-libraries-dev        12.4.0                        0    nvidia
cuda-nvrtc                12.4.127                      0    nvidia
cuda-nvrtc-dev            12.4.127                      0    nvidia
cuda-nvtx                 12.4.127                      0    nvidia
cuda-opencl               12.6.37                       0    nvidia
cuda-opencl-dev           12.6.37                       0    nvidia
cuda-profiler-api         12.6.37                       0    nvidia
cuda-runtime              12.4.0                        0    nvidia
cuda-version              12.6                          3    nvidia
expat                     2.6.2                hd77b12b_0
filelock                  3.13.1          py312haa95532_0
intel-openmp              2023.1.0         h59b6b97_46320
jinja2                    3.1.4           py312haa95532_0
libcublas                 12.4.2.65                     0    nvidia
libcublas-dev             12.4.2.65                     0    nvidia
libcufft                  11.2.0.44                     0    nvidia
libcufft-dev              11.2.0.44                     0    nvidia
libcurand                 10.3.7.37                     0    nvidia
libcurand-dev             10.3.7.37                     0    nvidia
libcusolver               11.6.0.99                     0    nvidia
libcusolver-dev           11.6.0.99                     0    nvidia
libcusparse               12.3.0.142                    0    nvidia
libcusparse-dev           12.3.0.142                    0    nvidia
libffi                    3.4.4                hd77b12b_1
libnpp                    12.2.5.2                      0    nvidia
libnpp-dev                12.2.5.2                      0    nvidia
libnvfatbin               12.6.20                       0    nvidia
libnvfatbin-dev           12.6.20                       0    nvidia
libnvjitlink              12.4.99                       0    nvidia
libnvjitlink-dev          12.4.99                       0    nvidia
libnvjpeg                 12.3.1.89                     0    nvidia
libnvjpeg-dev             12.3.1.89                     0    nvidia
libuv                     1.48.0               h827c3e9_0
markupsafe                2.1.3           py312h2bbff1b_0
mkl                       2023.1.0         h6b88ed4_46358
mkl-service               2.4.0           py312h2bbff1b_1
mkl_fft                   1.3.8           py312h2bbff1b_0
mkl_random                1.2.4           py312h59b6b97_0
mpmath                    1.3.0           py312haa95532_0
networkx                  3.3             py312haa95532_0
numpy                     1.26.4          py312hfd52020_0
numpy-base                1.26.4          py312h4dde369_0
openssl                   3.0.14               h827c3e9_0
pip                       24.2            py312haa95532_0
python                    3.12.4               h14ffc60_1
pytorch                   2.4.0           py3.12_cuda12.4_cudnn9_0    pytorch
pytorch-cuda              12.4                 h3fd98bf_6    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pyyaml                    6.0.1           py312h2bbff1b_0
setuptools                72.1.0          py312haa95532_0
sqlite                    3.45.3               h2bbff1b_0
sympy                     1.12            py312haa95532_0
tbb                       2021.8.0             h59b6b97_0
tk                        8.6.14               h0416ee5_0
typing_extensions         4.11.0          py312haa95532_0
tzdata                    2024a                h04d1e81_0
vc                        14.40                h2eaa2aa_0
vs2015_runtime            14.40.33807          h98bb1dd_0
wheel                     0.43.0          py312haa95532_0
xz                        5.4.6                h8cc25b3_1
yaml                      0.2.5                he774522_0
zlib                      1.2.13               h8cc25b3_1
```

**Comment by sdatkinson** [comment]
Almost forgot: another note on [my above comment](https://github.com/sdatkinson/neural-amp-modeler/issues/450#issuecomment-2293728211): I was able to reproduce the Issue with a fresh install of the `nam` environment. Reverting PyTorch to version 2.3.0 is a workaround that seems to at least prevent crashing.

**Comment by prash-p** [comment]
> One quick note on the OP:
> 
> > Re-enable/sync to the old parametric code.
> 
> Old versions aren't supported, so I'm not going to be trying to reproduce or fix on an old version. Once this works on the current version then I'll consider this resolved.

Quick question: on your vibrato capture on youtube did you use lstm, wavenet or something else entirely?

**Comment by satellitnorden** [comment]
> Old versions aren't supported, so I'm not going to be trying to reproduce or fix on an old version. Once this works on the current version then I'll consider this resolved.

Totally fair, I understand. (: Just confirmed this happens for me on regular snapshot models as well, so seems to be something wrong with the "base" LSTM model-

**Comment by sdatkinson** [comment]
Interesting: I've seen [elsewhere](https://github.com/pytorch/pytorch/issues/32564) that this can be (maybe?) caused by an out-of-memory.

I've observed this when breaking on https://github.com/sdatkinson/neural-amp-modeler/blob/6d33b41cf9f361b774e25c4e1e6340503ae7410e/nam/models/recurrent.py#L259:

![image](https://github.com/user-attachments/assets/57491e8e-f4bd-46bc-bd0e-17cca0423218)

...which is a very _coincidental_ number.

Funnily enough, the attempt to MWE above doesn't fail out if I use a similarly-sized input. So that's also mysterious to me. I've been stripping away parts to NAM's wrapper around PyTorch's LSTM and haven't quite gotten to whatever makes things "click".

Add to it that the input is indeed (apparently) contiguous...
![image](https://github.com/user-attachments/assets/b85a2da1-4e3f-4cfc-abca-c2044af11a30)

...and I think there's some funny business going on.

Two ways to fix this:
1. Catch the exception and process the input in chunks (eww, don't like that)
2. Continue hacking away to see if I can figure out exactly what makes things go south and see if it can be addressed.

Gonna pick 2 for a while yet.

**Comment by sdatkinson** [comment]
Aha, forgot that PyTorch's LSTM uses non-batch-first tensors as input by default.

This reproduces the error without any NAM code:

```python
import torch


def test_case(sequence_length: int):
    # device = "cpu"  # SUCCESS
    device = "cuda"  # FAIL
    print("Attempt...")
    # Not important what values are chosen for these
    batch_size = 2
    input_size = 4
    x = torch.randn(sequence_length, batch_size, input_size).to(device)
    model = torch.nn.LSTM(num_layers=3, input_size=input_size, hidden_size=24).to(
        device
    )
    model(x)


test_case(sequence_length=65535)  # Succeeds
try:
    test_case(sequence_length=65536)
    print("SUCCESS!")
except RuntimeError as e:
    print("FAIL!")
    print(e)
```

Output:
```
Attempt...
Attempt...
FAIL!
cuDNN error: CUDNN_STATUS_NOT_SUPPORTED. This error may appear if you passed in a non-contiguous input.
```

So I'm (surprisingly) betting this is a PyTorch bug. (I'm confident it's not a NAM bug.) I'll report it over there and see what I learn.

I also bet I can make a workaround for NAM in the meantime.

**Comment by sdatkinson** [comment]
We'll see what we hear: https://github.com/pytorch/pytorch/issues/133751

**Comment by sdatkinson** [comment]
Should work on `main` now. Lmk if not

**Comment by sdatkinson** [comment]
@lbdhr the fix should work for 2.5.1. Open me a new Issue and include the stack trace you're getting?

**Comment by lbdhr** [comment]
> @lbdhr the fix should work for 2.5.1. Open me a new Issue and include the stack trace you're getting?

Thank you very much for your prompt response, @sdatkinson . Please see the new issue here: https://github.com/sdatkinson/neural-amp-modeler/issues/512
